Version 1.5.1

pauljamesdfe opened the pull request:
This is a updated version of the Local skills dashboard. Changes were made based on user feedback. The major changes are:
1) Refresh of the summary page
2) Remove LEP maps and add in National picture
3) Add in page wide subgroup filters to the Local Skills data page

QA
1) Refresh of the summary page
a) Check the page works as intended. Does it make sense? Does it flow? Does it work for all areas (I have already tested this but some spot tests won't hurt)? Does it offer the right level of information?
b) Code changes are mostly focused in R/summaryTab.r (all new code) and server.R (some new sections of code and some altered sections of code.

2) Remove LEP maps and add in National picture
a) Check across the dashboard that there is no LEP and that the National option is present
b) Check the National option works on the Summary and Local skills tab. I have gone through all metrics on the Local skills tab but spot checks will confirm.
c) Code changes will be mostly across the file within /importData where LEP data is no longer imported and National data is. 
d) There is a new map for the national option on the Local skills tab. Check this works and check the import code for the boundary data (/importData/importBoundaries.R). When this option is selected the bottom right map shows all LAs - sense check is this useful or too much?

3) Add in page wide subgroup filters to the Local Skills data page
a) Check this is intuitive for a user. Check it works.
b) Visual code, Code changes to allow this to be seen will be in ui.R and in server.R. Some of the filters interact (eg the subgroup filter and those above the breakdown chart) - check this is working. 
c) import code. For job ads now all years data is imported to show the time series of subgroups /importData/importJobAds.R. For others the structure of the data has been changed to allow this functionality to work.

Miscellaneous
Some of the downloads were not working (this is a legacy problem I only just noticed). I have corrected these mostly in server.R but also in importData/combineData.R

pauljamesdfe commented:
@jlwalker5523 Thanks for going through this. I've made some changes to cover some (but not all) of your points: 
> Comments/issues:
> 
> * Businesses section: data commentary describes change in business count as 'fallen' when there is 0% difference, and shall we include text to explain these are the most over and under represented industries versus national average.
------Here I have added a decimal place to the % value so you can see the smaller changes. I've changed the text a bit to reflect that it is against the England average. -----
> * Shall we add the explanatory text for the previous year comparison (it's particularly less obvious for the online job ads).
-----I think in your version the text values above the charts had disappeared. They are now back on. Let me know if that makes it clearer. -----
> * England summary: employment rate showing as red between 2023 and 2024 for the same %, and no chart is appearing under the Businesses section for me.
---- As above I added a decimal point so you can see this change now. The charts do not show here because the charts are a comparison to England - unless there was something different you thought would be helpful here just for England  ------
> * Data call outs appear in front of the charts for Brighton and Hove LSIP so it's difficult to see which data point you are viewing. I can not find this for any other locations so possibly an isolated issue.
---- I think this is just because the name is so long. I don't know if I can do anything about this sadly -----

Can you have a look please and see if you are satisfied. @jvevansdfe can you have a look at what I've done also to see if you agree. I haven't always covered all of Jody's excellent points because in some cases I can't! But hoping it's still ok to go forward?

jvevansdfe commented:
Thanks both! On @pauljamesdfe's responses to @jlwalker5523's QA of the summary page:
- Business section: I agree with the changes. Please could you add two further changes:
  - If the rounded figure is 0.0, change the text to 'remained broadly stable (less than 0.1% change)' to cover all bases
  - Change the second half of the England comparison statement in line with the first (e.g. '...and a lower proportion of X businesses'
- Explanatory text: I'm happy with this but interested in @jlwalker5523's views / any suggestions to make clearer
- England summary: would it be possible to make grey if the change rounds to 0.0? I'm fine with no chart in the business section.
- Data call outs: agree we will just have to live with this.

I have also completed a further QA myself across the dashboard, this included:
- Overarching review of dashboard (run locally), focusing on areas affected by code changes
- Testing of all new features
- Spot checks of new data (breakdowns) against source data, for a selection of metrics and breakdowns
- Testing of downloads feature

**I have spotted a few additional areas to address:**

Skills Imperative data: 
- The hover-over label in the timeseries graph reads 'projected employment growth from 2023 to 2035', but the datapoint represents the year on year change.
- Historic issue: we compare to 2023. Is there a reason for this or can we update to 2024?

All skills data pages:
- If you select a different breakdown in the top menu, the graph in the bottom left will update to show that breakdown, but the heading for that graph will remain what it was before (so may refer to a different breakdown) 
- In the timeseries graph there is often a lot of suppression if viewing a breakdown for an LA. Sometimes there will be a data point at one time point, with no data points either side. In these cases you can see the data point label if you hover over the graph in its location, but there is nothing visible on the graph (as the data points only show up when they connect (example: Amber Valley > Employment > 12 Other managers and proprietors breakdown - there is a data point for 2024/25 but only the connected 21/22 and 22/23 show up in the graph). I am happy to put this on the backlog if there is no quick fix.

Data downloads (these seem to be historic issues):
- I'm not seeing any values for the 2020/21 time period when selecting employment volume with occupation (SOC Sub-Major Group) as the breakdown 
- Some of the options in the data menu are variable names (e.g. acheivementAims) instead of plain text descriptions (i.e. Achievement aims) - is it possible to correct?
- Some options in the data menu (e.g. economicallyactiveRate) have a selectable option of 'breakdown' in the breakdown menu

pauljamesdfe commented on importData/importEnterprises.R:
```diff
@@ -1,24 +1,24 @@
   bind_rows(
     # user defined pt 1 (do in two parts as falls over all together)
     nomisr::nomis_get_data("NM_142_1",
-                   geography =
```
I think it might just be that the new files are indented before "geography" and before "industry"

jlwalker5523 commented on importData/importBoundaries.R:
@pauljamesdfe I have ran this code line by line and sense checked the changes. Is there any particular checks I can do on the new England boundary?

jlwalker5523 replied:
@pauljamesdfe flagging this comment in case it was missed.

jlwalker5523 commented on importData/importFeAims.R:
```diff
@@ -7,22 +7,20 @@ I_FeProvLevelAge <- read.csv(file = paste0("./Data/", folder, "/", list.files(pa
 F_FeProvLevelAge <- I_FeProvLevelAge %>%
   # filter uneeded columns and rows
```
@pauljamesdfe small typo in code annotations

pauljamesdfe replied:
Changed https://github.com/dfe-analytical-services/lsip_dashboard/pull/469/commits/54701afc2fa76f02e8d1d90a5891cd9bd36a7682

jvevansdfe commented on server.R:
```diff
@@ -460,19 +460,19 @@ server <- function(input, output, session) {
         paste0(
           ", while across England businesses have ",
-          ifelse(englandChange > 0, "grown ", "fallen "),
+          ifelse(englandChange > 0.0005, "grown ", ifelse(englandChange < 0.0005, "fallen ", "remained broadly stable (less than 0.1% change)")),
```
As above: this logic needs reviewing - currently anything less than 0.05% is labelled as 'fallen' and 'broadly stable' is only applied to changes of exactly 0.05%. Also the percentage change will still be added after the broadly stable statement, which is not necessary.

pauljamesdfe replied:
Fixed i think (https://github.com/dfe-analytical-services/lsip_dashboard/pull/469/commits/c05997cedb204636324cd7336f559d2fd5c44f32)

jvevansdfe replied:
Thanks, looks good now

pauljamesdfe commented:
@jvevansdfe  re https://github.com/dfe-analytical-services/lsip_dashboard/pull/469#issuecomment-3132898559: 
Changed skills imperative to 2024 to 2035. 
Changed the hover on the time chart to Annual change. 
Commit: https://github.com/dfe-analytical-services/lsip_dashboard/pull/469/commits/6aa82e5078f53fada073e49a55d965f4aa5f2178

pauljamesdfe commented:
@jvevansdfe re https://github.com/dfe-analytical-services/lsip_dashboard/pull/469#issuecomment-3132898559:

Made breakdown title and comment update with top breakdown filter
Commit: https://github.com/dfe-analytical-services/lsip_dashboard/pull/469/commits/8584d58467e44e00cf9f63dd746c09c4c543612e

pauljamesdfe commented:
@jvevansdfe re https://github.com/dfe-analytical-services/lsip_dashboard/pull/469#issuecomment-3132898559:

Added dots to the line charts so you can see singular, unconnected points
Commit: https://github.com/dfe-analytical-services/lsip_dashboard/pull/469/commits/c0232fe57ef7f66167c189e02b26d487d85520e1

jlwalker5523 commented on importData/importNomis.R:
QA:
- Code runs without errors
- Minor code changes: sense checked removal of LEP code

jlwalker5523 commented on importData/importEmp.R:
QA:
- Code runs without errors

jlwalker5523 commented on importData/importEmpOcc.R:
QA:
- Code runs without errors
- No code changes to QA

jlwalker5523 commented on importData/importQualification.R:
QA:
- Code runs without errors
- Checked valueText col is working as expected (returns "c" for 0, otherwise value).

jlwalker5523 commented on importData/importFeSsa.R:
QA:
- Code runs without errors
- No further QA as code matches original.

pauljamesdfe commented:
@jvevansdfe re https://github.com/dfe-analytical-services/lsip_dashboard/pull/469#issuecomment-3132898559:

>I'm not seeing any values for the 2020/21 time period when selecting employment volume with occupation (SOC Sub-Major Group) as the breakdown
There is no submajor SOC2020 before 21/22 (there is SOC2010 so I guess we could use this, but not sure about the alignment). I suggest leaving as is. See here for the data https://www.nomisweb.co.uk/livelinks/18571.xlsx

jlwalker5523 commented on importData/importDestinations.R:
QA
- Code runs without errors
- Checked new national data matches source (nrows and % sustained destinations)
- Checked filtering (line 20-22) is working as expected by filtering original df
- Checked totals match original dfs after binding

jlwalker5523 commented on importData/importFeAims.R:
QA:
- Code runs without errors 
- Sense checked removal of LEP related code

pauljamesdfe commented:
@jvevansdfe re https://github.com/dfe-analytical-services/lsip_dashboard/pull/469#issuecomment-3132898559:

> Some of the options in the data menu are variable names (e.g. acheivementAims) instead of plain text descriptions (i.e. Achievement aims) - is it possible to correct?
> Some options in the data menu (e.g. economicallyactiveRate) have a selectable option of 'breakdown' in the breakdown menu

I've corrected achievement aims and another aims one, and got rid of some of the metrics not used in the dashboard (that includes the one with the Breakdown problem).

Commit: https://github.com/dfe-analytical-services/lsip_dashboard/pull/469/commits/fcfdb88ec136d2bbbcaba4f72557432e7f6612c1

jlwalker5523 commented on importData/importDestinations.R:
```diff
@@ -30,24 +37,18 @@ destinationsWithAreas <-
-  bind_rows(destinationsWithAreas)#%>%
-
-groupedStats <- destinationsEngland %>%
+groupedStats <- destinationsWithAreas %>%
```
@pauljamesdfe there is another df called groupedStats for business count, not sure if it's intentional but destinations data is overwritten when all the code is run.

pauljamesdfe commented on importData/combineData.R:
```diff
@@ -38,24 +38,28 @@ C_Geog <- neatGeog %>%
   left_join(
     (C_localSkillsDataset %>%
        filter(
-         breakdown == "Total", latest == 1,
```
Yeah we used to just show the totals (no subgroups) in the maps, now we need that subgroup data still in so we can show it on the maps if it is chosen.

jlwalker5523 commented:
**QA
3. Add in page wide subgroup filters to the Local Skills data page**

Checks:
- Tested that the data is updating with subgroup filters (spot checked job ads as below, and business count and % for 2 subgroups against source data). 
- Tested subgroup data and breakdown update when used together.
- Checked import job ads - latest data filter removed, data across years showing on dashboard, spot checked previous years data against source (one SOC sub-major group across two geographies). 

Comments/issues: 
- For business count, when split by subgroup the text commentary for the chart comparing businesses by size pulls out the difference between large business % - is this intentional as some of the differences are larger across other groups?
- As an overall comment, its quite difficult to QA some parts of the code due to complexity and/or nesting so creating new variables before joining and binding would make it easier to QA and including more annotations throughout to breakdown each step (e.g., in combineData.R). 

Note: combineData.R is yet to be QA'd.

pauljamesdfe commented on ui.R:
```diff
@@ -485,14 +479,22 @@ div.myspecial-popup div.leaflet-popup-content-wrapper {
+              tags$li("Remove LEP maps and add in National picture."),
+              tags$li("Add in page wide subgroup filters to the Local Skills data page."),
+              tags$li("Update Skills Imperative projection long term growth metric to 2024 to 2035."),
+              tags$li("Utilise the populations published in FE data to calculate grouped FE per 10k statistics."),
```
Changed: https://github.com/dfe-analytical-services/lsip_dashboard/pull/469/commits/e1998122194d48e30b184e9cb692f704e6e7dc22

pauljamesdfe commented:
Thanks @jlwalker5523. Some fixes and comments on these:
> Other comments/issues:
> 
> * The chart axis on the local skills data page doesn't look to be working correctly e.g., for national online job ads 1.5m appears at 2m on the axis. Plus, the same data point appears differently (assume this because the underlying % is slightly different e.g., employment rate for the last 3 years. Zooming out on the scale might help resolve small changes appearing large too.

Yeah this isn't great, but not so easy to fix. We use the scales::cut_short_scale function to get whether it is thousands, millions etc then it adds the k, M after. We could write our own function though that would improve this I think. @jvevansdfe One for the backlog?

> * When looking at an LSIP or MCA I get an error if I select England as the comparison area (Error: factor level [3] is duplicated). Comparison works if select national and the LSIP/MCA/LA as comparison area so may need to tell users it only works that way around.

I've fixed this I hope. I've made it so England is always grey when chosen as a comparator so it matches the breakdown. Could you check it's doing what you thought it should please. Commit here: https://github.com/dfe-analytical-services/lsip_dashboard/pull/469/commits/2eafebbd6f736259d22e6a12b43b74513d2a47cb

> * LA map: I think it's useful to see hotspots and how the metric varies over England but it  isvisually very detailed (might be good to see how users find it?). I think we could make it more useful if we included the top and bottom % in the chart commentary to show the range as I find myself looking for the LA on the map.

Yeah I thought the same that it might be a bit much. Any ideas from you or @jvevansdfe would be helpful. I think adding the values is good idea - I'll try and do that if I get time otherwise one for the backlog.

> * Quick one: noticed that a date is appearing on the bottom left hand side when there is no chart e.g., on employment rate.
I can't see this one? Can you screenshot please.

jvevansdfe commented:
@pauljamesdfe and @jlwalker5523 re https://github.com/dfe-analytical-services/lsip_dashboard/pull/469#issuecomment-3140531387 and https://github.com/dfe-analytical-services/lsip_dashboard/pull/469#issuecomment-3140549015

Thanks both. On potential issues for the backlog:

- Axis scales: unfortunately I think we do need to fix this as for the England volume charts there are some that just look wrong (e.g. business count has a 2M value for every label on the Y axis). I am happy for this to be a quick work around for now though. For example adding an extra if statement to capture graphs where England is the geography choice, and changing the accuracy to 0.01 (which would mean on the business count graph as an example, we would have labels like 2.37M, 2.38M etc). Not the nicest solution but we can then take our time with a better one.
- LA map: Yes I'll add this to the backlog. I think this fits with some of our longer term plans to consider more narrative and summary information for the dashboard. I think it is ok as it is for now as you can zoom in and out to see the detail and most users will be interested in a particular area of the country.
- Code structure: Yes I'll add this to the backlog. It's an important point to raise but a big job to tackle so we should do that separately to this update.

jlwalker5523 reviewed (approved):
Thanks - I think that's much clearer!

QA: confirmed changes to decimal points and text appears in dashboard.

pauljamesdfe commented:
Hi @jlwalker5523. Sorry an additional bit to QA. I noticed the links on the Summary tab weren't working (code wasn't carried over during the merge). I've fixed that here: 
Commit: https://github.com/dfe-analytical-services/lsip_dashboard/pull/469/commits/fed1e912defd36a1188922862540e03967b5ef9b
Thanks

jlwalker5523 reviewed (commented):
Thanks Paul! I have sense checked the added boxes matches the previous dashboard and charts. 

@jvevansdfe please can you see if this needs any extra checks as it's Rshiny code.

jlwalker5523 commented:
> > * The chart axis on the local skills data page doesn't look to be working correctly e.g., for national online job ads 1.5m appears at 2m on the axis. Plus, the same data point appears differently (assume this because the underlying % is slightly different e.g., employment rate for the last 3 years. Zooming out on the scale might help resolve small changes appearing large too.
> 
> Yeah this isn't great, but not so easy to fix. We use the scales::cut_short_scale function to get whether it is thousands, millions etc then it adds the k, M after. We could write our own function though that would improve this I think. @jvevansdfe One for the backlog?

Ah I see! Do we know why it returns 2M for values starting with 1M? I am thinking if people use the axis and not the data labels it could lead people to the wrong conclusions e.g., it looks as though vacancies were above 2M between July 21 and Nov 22. 

> > * When looking at an LSIP or MCA I get an error if I select England as the comparison area (Error: factor level [3] is duplicated). Comparison works if select national and the LSIP/MCA/LA as comparison area so may need to tell users it only works that way around.
> 
> I've fixed this I hope. I've made it so England is always grey when chosen as a comparator so it matches the breakdown. Could you check it's doing what you thought it should please. Commit here: [2eafebb](https://github.com/dfe-analytical-services/lsip_dashboard/commit/2eafebbd6f736259d22e6a12b43b74513d2a47cb)

Thanks Paul - yep I can now select England as the comparator, with an LSIP/MCA as the main selection. 

> > * Quick one: noticed that a date is appearing on the bottom left hand side when there is no chart e.g., on employment rate.
> >   I can't see this one? Can you screenshot please.

Not a big deal at all! E.g., on employment rate: 
<img width="781" height="545" alt="image" src="https://github.com/user-attachments/assets/5b8bd745-ee81-4b5f-a37e-a3a90c44c5c8" />

jlwalker5523 commented:
> Hi @jlwalker5523. Sorry an additional bit to QA. I noticed the links on the Summary tab weren't working (code wasn't carried over during the merge). I've fixed that here: Commit: [fed1e91](https://github.com/dfe-analytical-services/lsip_dashboard/commit/fed1e912defd36a1188922862540e03967b5ef9b) Thanks

@pauljamesdfe checked all the links were working and take me to the relevant metric on local skills data.

jlwalker5523 commented:
> @jlwalker5523 @jvevansdfe
> 
> > * Axis scales: unfortunately I think we do need to fix this as for the England volume charts there are some that just look wrong (e.g. business count has a 2M value for every label on the Y axis). I am happy for this to be a quick work around for now though. For example adding an extra if statement to capture graphs where England is the geography choice, and changing the accuracy to 0.01 (which would mean on the business count graph as an example, we would have labels like 2.37M, 2.38M etc). Not the nicest solution but we can then take our time with a better one.
> 
> I've changed this so if the chart goes into millions it is 2 dp, so x.xxM [2659c49](https://github.com/dfe-analytical-services/lsip_dashboard/commit/2659c49ce531df79b8a9a3bf09fd164193c5accc)

@pauljamesdfe changes look good to me, thank you! @jvevansdfe please can you sense check the server code too.

jvevansdfe commented:
@pauljamesdfe re https://github.com/dfe-analytical-services/lsip_dashboard/pull/469#issuecomment-3143831259
> > * Axis scales: unfortunately I think we do need to fix this as for the England volume charts there are some that just look wrong (e.g. business count has a 2M value for every label on the Y axis). I am happy for this to be a quick work around for now though. For example adding an extra if statement to capture graphs where England is the geography choice, and changing the accuracy to 0.01 (which would mean on the business count graph as an example, we would have labels like 2.37M, 2.38M etc). Not the nicest solution but we can then take our time with a better one.
> 
> I've changed this so if the chart goes into millions it is 2 dp, so x.xxM [2659c49](https://github.com/dfe-analytical-services/lsip_dashboard/commit/2659c49ce531df79b8a9a3bf09fd164193c5accc)

Sorry I've realised this also applies to some charts that are in thousands, because the problem is to do with the difference between the min and max of the data.  I've checked a selection of charts and they all seem to have 6 axis labels or fewer, so the problem should only be occurring when the difference between min and max is less than 6 units (units being either K or M - whatever is displayed). In these cases you'll get at least one repeated label if the chart does have the max of 6 labels (I know this is dynamic, so won't always be the case). Could we change the logic to capture that? Another option is just apply the 0.01 precision to everything until we figure out a nicer fix.

pauljamesdfe commented:
@jlwalker5523 Thanks! I had a method that turns out was only working half the time. Hopefully fixed now here: 
https://github.com/dfe-analytical-services/lsip_dashboard/pull/469/commits/29d5cf48f05e8197e088e8bbfc8f553147d60e7e

> > > * Quick one: noticed that a date is appearing on the bottom left hand side when there is no chart e.g., on employment rate.
> > >   I can't see this one? Can you screenshot please.
> 
> Not a big deal at all! E.g., on employment rate: <img alt="image" width="781" height="545" src="https://private-user-images.githubusercontent.com/204232436/473399590-5b8bd745-ee81-4b5f-a37e-a3a90c44c5c8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTQwNTg2MzEsIm5iZiI6MTc1NDA1ODMzMSwicGF0aCI6Ii8yMDQyMzI0MzYvNDczMzk5NTkwLTViOGJkNzQ1LWVlODEtNGI1Zi1hMzdlLWEzYTkwYzQ0YzVjOC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwODAxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDgwMVQxNDI1MzFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00MTdjNzZmYzk1NWMyMDA0YjQ0YTE5YWRmYzdjM2U5MmQxMzAzOGM3ZjQ4Y2ZkNmE5OWNjZDFmZmQwYjQ2NzQ5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.ODi7HMzPOcraj_rf2OLL5KdlY4aZ8EEUtaYo_S_0mwA">

pauljamesdfe commented:
Thanks @jlwalker5523. Employment rate is the only metric that does have breakdowns of all the employment metrics (ie Unemployment volumes do not have breakdowns so i wouldn't want to direct people there). So just as a hint to get any employment breakdowns I signpost people there. Happy to change the text if you think it could be clearer.
> @pauljamesdfe this has fixed the a date showing, thank you! But I have spotted the text which appears isn't updating with the metric. E.g., employment rate correctly shows > Employment rate currently has no breakdowns. Switch to **Employment metric** for occupation and industry breakdowns. Self-employment/unemployment/inactive rate show this exact text too.

jvevansdfe commented on importData/importSkillsImp.R:
```diff
@@ -174,20 +162,20 @@ empGrowth <- employmentProjections %>%
-  filter(chartPeriod != 2023) %>%
-  mutate(metric = "employmentProjectionGrowth2023to2035")
+  filter(chartPeriod != 2024) %>%
+  mutate(metric = "employmentProjectionGrowth2024to2035")
```
Thanks looks good.
